pkg/services/ngalert/state/state_tracker.go

```diff
@@ -136,57 +134,120 @@ func (st *StateTracker) ProcessEvalResults(alertRule *ngModels.AlertRule, result
+	switch a.State {
+	case eval.Alerting:
+		if !(alertRule.For > 0) {
+			// If there is not For set, we will set EndsAt to be twice the evaluation interval
+			// to avoid flapping with every evaluation
+			a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+			return a
+		}
+		a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+	case eval.Pending:
+		if result.EvaluatedAt.Sub(a.StartsAt) > alertRule.For {
+			a.State = eval.Alerting
+			a.StartsAt = result.EvaluatedAt
+			a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+			a.Annotations["alerting_at"] = result.EvaluatedAt.String()
 		}
-		st.set(currentState)
-		return currentState
-	case currentState.State == eval.Normal && result.State == eval.Alerting:
-		st.Log.Debug("state transition from normal to alerting", "cacheId", currentState.CacheId)
-		currentState.State = eval.Alerting
-		currentState.LastEvaluationTime = result.EvaluatedAt
-		currentState.StartsAt = result.EvaluatedAt
-		currentState.EndsAt = result.EvaluatedAt.Add(alertRule.For * time.Second)
-		currentState.EvaluationDuration = evaluationDuration
-		currentState.Results = append(currentState.Results, StateEvaluation{
-			EvaluationTime:  result.EvaluatedAt,
-			EvaluationState: result.State,
-		})
-		currentState.Annotations["alerting"] = result.EvaluatedAt.String()
-		st.set(currentState)
-		return currentState
-	case currentState.State == eval.Alerting && result.State == eval.Normal:
-		st.Log.Debug("state transition from alerting to normal", "cacheId", currentState.CacheId)
-		currentState.State = eval.Normal
-		currentState.LastEvaluationTime = result.EvaluatedAt
-		currentState.EndsAt = result.EvaluatedAt
-		currentState.EvaluationDuration = evaluationDuration
-		currentState.Results = append(currentState.Results, StateEvaluation{
-			EvaluationTime:  result.EvaluatedAt,
-			EvaluationState: result.State,
-		})
-		st.set(currentState)
-		return currentState
 	default:
-		return currentState
+		a.StartsAt = result.EvaluatedAt
+		if !(alertRule.For > 0) {
+			a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+			a.State = eval.Alerting
+			a.Annotations["alerting_at"] = result.EvaluatedAt.String()
+		} else {
+			a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+			if result.EvaluatedAt.Sub(a.StartsAt) > alertRule.For {
+				a.State = eval.Alerting
+				a.Annotations["alerting_at"] = result.EvaluatedAt.String()
+			} else {
+				a.State = eval.Pending
+			}
+		}
+	}
+	return a
+}
+
+func (a AlertState) resultError(alertRule *ngModels.AlertRule, result eval.Result) AlertState {
+	if a.StartsAt.IsZero() {
+		a.StartsAt = result.EvaluatedAt
+	}
+	if !(alertRule.For > 0) {
+		a.EndsAt = result.EvaluatedAt.Add(time.Duration(alertRule.IntervalSeconds*2) * time.Second)
+	} else {
+		a.EndsAt = result.EvaluatedAt.Add(alertRule.For)
+	}
+	if a.State != eval.Error {
+		a.Annotations["last_error"] = result.EvaluatedAt.String()
```

Comment:
Is this correct? I'm just curious about whenever we meant to sent last_error as timestamp here.

I'm a bit puzzled by the inclusion of these annotations - they might be correct, I just want to get an understanding of why.